### ISSUE TITLE
`single_range_in_vec_init` stronger than expected, when type is known

### ISSUE THREAD

**Issue by matzemathics** [rust-lang/rust-clippy#11086]
### Summary

The `single_range_in_vec_init` warning seems a bit overly cautious, if the type resulting from an invocation of a `vec!` macro is already known "from the outside", e.g. via a type annotation or when it is used as an argument / return value.

For example
```rust
let foo: Vec<Range<usize>> = vec![1..5]; // causes "warning: a `Vec` of `Range` that is only one element"
```
In this case clippy might infer, that the user indeed wanted a `Vec<Range<...>>`, because otherwise the program could not even have type checked. The same is true for this example:

```rust
fn barify(v: Vec<Range<usize>>) {
  ...
}

let v = vec![3..42]; // causes "warning: a `Vec` of `Range` that is only one element"
barify(v);
```
One explicit way to remedy this is using explicit lengths:
```rust
let foo = vec![1..5; 1]; // OK
```
although I am slightly worried that tools like cargo-fmt might at some point opt to remove that.

### Lint Name

single_range_in_vec_init

### Reproducer

I tried this code:

```rust
use std::ops::Range;

fn main() {
    let _foo: Vec<Range<usize>> = vec![1..5];
}
```

I saw this happen:

```
warning: a `Vec` of `Range` that is only one element
 --> src/main.rs:4:35
  |
4 |     let _foo: Vec<Range<usize>> = vec![1..5];
  |                                   ^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_range_in_vec_init
  = note: `#[warn(clippy::single_range_in_vec_init)]` on by default
help: if you wanted a `Vec` that contains the entire range, try
  |
4 |     let _foo: Vec<Range<usize>> = (1..5).collect::<std::vec::Vec<usize>>();
  |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: if you wanted a `Vec` of len 5, try
  |
4 |     let _foo: Vec<Range<usize>> = vec![1; 5];
  |                                        ~~~~

warning: `single_range` (bin "single_range") generated 1 warning
```

I expected to see this happen:
No warnings produced.

**Comment by Centri3** [comment]
Ignoring it for arguments when it's a local would likely be pretty tricky, though type annotations should definitely be checked

**Comment by ArthurCose** [comment]
Something not currently covered by the PR:

```rust
use std::ops::Range;

struct A {
    ranges: Vec<Range<usize>>,
}

struct B(Vec<Range<usize>>);

fn main() {
    A { ranges: vec![0..5] };
    B(vec![0..5]);
}
```

The correct type of a struct's field is also known "from the outside".

**Comment by Centri3** [comment]
Should be simple enough to fix

**Comment by Centri3** [comment]
It ignores `ExprField`s now. This also means something like `Vec<&dyn Any>` will be skipped as well but I think due to the rarity of both that and passing a `Range<usize>` to it, it should be fine. (It's also debatable whether that should even be linted.)

**Comment by djc** [comment]
I'm also running into this in Quinn, in the [tests](https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/range_set/btree_range_set.rs#L323) for the `RangeSet::replace()` method. This definitely seems like a false positive.

**Comment by avantgardnerio** [comment]
This is coming up with a lot of false positives for our project as well. I'm concerned that it is valid to have/want a Vec with a single Range, and that the suggestions to correct it would alter functionality. If anyone goes and applies them blindly, they might cause subtle bugs.

**Comment by Centri3** [comment]
The help message should probably be altered to make sure this doesn't happen, since right now it acts as though this is almost certainly what they want

**Comment by synek317** [comment]
It also causes quite annoying issues in my project. I completely don't understand the intention behind this lint and why is it enabled by default.

**Comment by bugadani** [comment]
This lint doesn't tell me what to do if I actually WANT an array/vec of ranges, which makes its strength even more bothersome.

**Comment by tisonkun** [comment]
+1 I have a function:

```rust
pub async fn read_ranges(
    &self,
    loc: Location,
    ranges: Vec<Range<u64>>,
) { ... }
```

.. and a function call:

```rust
storage
  .read_ranges(
      loc,
      vec![start as u64..end as u64],
  )
  .await
```

The auto fix would fail compilation:

```rust
storage
  .read_ranges(
      loc,
      (start as u64..end as u64).collect::<std::vec::Vec<u64>>(), // Vec<u64>, while surely we need Vec<Range<u64>>; the content has been wrongly changed anyway.
  )
  .await
```

**Comment by AmmarAbouZor** [comment]
I don't understand why this lint is enabled by default too. If you use `vec![0..10]` where your code expecting a `Vec<usize>` then you'll get a compiler error anyway. Rust is statically typed language, and the compiler is responsible for such mistakes not the linter. 

Lints producing such fix hints shouldn't be enabled by default
```sh
help: if you wanted a `Vec` that contains the entire range, try
   |
39 -     let rngs: Vec<Range<usize>> = vec![1..10];
39 +     let rngs: Vec<Range<usize>> = (1..10).collect::<std::vec::Vec<usize>>();
   |
help: if you wanted a `Vec` of len 10, try
   |
39 -     let rngs: Vec<Range<usize>> = vec![1..10];
39 +     let rngs: Vec<Range<usize>> = vec![1; 10];
   |
```

**Comment by williamwutq** [comment]
I wonder if it is possible to make clippy not produce this warning when the type is known to be `Vec` or similar. I my case, I have a function:

```rust
pub async fn read_batched<I>(&self, ranges: I) -> io::Result<Vec<Vec<u8>>>
where
    I: IntoIterator<Item = std::ops::Range<u64>>,
```

Where both of its suggestions would fail to compile. The correct solution here is actually using iterator: `get_batched(core::iter::once(0..10))`

I don't know if it is possible for clippy to know the precise type expected from function calls, but not issuing a spurious warning when the type is a collection of ranges would be nice.

**Comment by netizen-a** [comment]
I'm also getting a false positive on this clippy warning. I designed this API to take an array, so not taking a Vec is intended for this:
```
let two = two.slice([0..out_features])?;
```
but clippy is pushing me to use Vec anyway despite the API saying it's not asking for a Vec, which is blatantly impossible.
